Unity-Technologies/ScriptableRenderPipeline
# Fix and cleanup pass on HDRP master node (Unlit, Fabric, lit, stacklit, PBR)

### Purpose of this PR
There is a lot of inconsistency between all shader graph in HD: Unlit, Fabric, lit, stacklit, PBR. Goal of this PR is to fix various reported bug and clean the code.

What is not the goal: We don't factor the code in the PR. This is a pity as many many code is duplicated but we have been agree that we don't touch anything and just copy/paste until the refactor of the API

What this PR do:

clean:
- Share BuildWorldToTangent functions
- Clean unused keyword / define
- Fix indentation
- clean PBR subshader code
- clean unused code in fabric

Decal:
- Replace _DECALS by _DISABLE_DECALS as for regular shader
- Move decal before specular AA
- add if (_EnableDecals) for decal test in all shader
- PBR master node: add decal support
- factor naming of disable SSR and decal in stacklit
- Use DisableDecals instead of Decals

SSR:
- update SSRdisable stencil test (not working in forward)
- Fix disable SSR in forward for shader graph and regular shader

General:
- update BuildSurfaceData prototype
- Rename HD Lit master to Lit Master
- fixed distortion blend mode in HD Lit
- fix distortion blend mode in stacklit
- Add everything to prepare distortion support on Unlit
- optimize a bit generated pass
- fix PBR sub shader stencil test
- fix PBR shader multicompile (Light layers and decals)
- Fix RequiresSplitLighting in Fabric master node
- update stencil state in Fabric
- remove two sided from unlit code (useless)
- add selection pass to unlit
- Fix bent normal for Fabric shader

---
### Release Notes
Please add any useful notes about the feature/fix that might be helpful for others.

---
### Testing status
**Katana Tests**:  https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=hdrp-clean-hd-master-node&automation-tools_branch=add-platform-filter&unity_branch=trunk

Katana is green.

**Manual Tests**:  Launch test locally

**Automated Tests**: What did you setup?

I prepare an automated test in a separate PR as I want to backport this PR to 2018.3 but not the associated test.


For QA  here is a checklist that is valid in general for shader graph (and will add in separate test).

- check if distortion working correctly with the correct blend mode for lit and unlit ?
- check that normal buffer correctly write smoothness and normal in forward + with depth prepass and motion vector pass
- check that SSR work with all material and SSR disable flags work in deferred and forward
- check that SSS work with all shader graph in deferred and forward
- check that decal work with all material and decal disable flags work in deferred and forward
- check how we output normalbuffer in case of coat normal etc.. in all shader (stacklit, fabric..)

---
### Overall Product Risks
**Technical Risk**: None, Low, Medium, High?

**Halo Effect**: None, Low, Medium, High?

---
### Comments to reviewers
There still work to do to fully update PBR and unlit to API used in HDLitMaster node
